File: tests/unit/optimizer/test_tvlqr_gains.py
"""Characterization tests for the chunked TVLQR feedback gains (trajOpt's K output).

trajOpt computes time-varying LQR tracking gains via compute_gains_chunked: it
runs the LQR backward pass over overlapping windows (tvlqr_len / tvlqr_overlap
seconds, stepping at the plan dt) and stitches them, dropping the overlap. This
is the design from McKeen (2025), Sec. 7.4.1 / Fig. 7.4: each chunk's gains
minimize the *remaining-chunk* cost (eq. 7.39), so the chunked gains are a
deliberate local-window approximation -- NOT expected to equal a single
full-horizon TVLQR. These tests therefore pin (a) structural validity, (b) the
no-op limit (a window covering the whole horizon reproduces the un-chunked
gains), (c) that shorter windows actually change the gains while staying
bounded, and (d) the gains' actual job: applying eq. 7.39 to a perturbed
closed-loop rollout tracks the plan far better than open loop.

These run in CI (the file name is not "test_alilqr", which CI deselects). The
trajectories are short (~20 steps) so the solves are quick, and every assertion
is structural or relative -- no absolute convergence thresholds.
"""
import sys
from pathlib import Path

import numpy as np
import pytest

ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(ROOT / "build"))
sys.path.insert(0, str(ROOT / "tests" / "debug" / "optimizer" / "configs"))

import saltro_py
from sat_0_3_rw import create_satellite as create_satellite_rw

SEC_PER_CENTURY = 36525.0 * 86400.0

# A window longer than the (200 s) trajectories below, with zero overlap, yields
# a single chunk spanning the whole horizon -- i.e. chunking is a no-op. (A
# nonzero overlap re-chunks even past the horizon, because the loop's next start
# = end + 1 - overlap_steps lands back inside the trajectory; and a window/dt
# ratio beyond INT_MAX overflows the step count, so keep the magnitude modest.)
SINGLE_CHUNK_LEN = 1000.0


def _make_settings(dt_seconds, tvlqr_len, tvlqr_overlap):
    ps = saltro_py.PlannerSettings()
    ps.init_traj.initcontroller = 2
    ps.num_passes = 1
    ps.passes[0].dt = dt_seconds
    ps.passes[0].ilqr.cost_tol = 1e-5
    ps.passes[0].ilqr.max_iters = 20
    ps.passes[0].auglag.max_outer_iters = 10
    ps.passes[0].auglag.constraint_tol = 1e-3
    c = ps.passes[0].cost
    c.angle = 1.0
    c.ang_vel = 1e1
    c.control_mult = 1.0
    c.mtq_control_weight = 1e-2
    c.rw_control_weight = 1.0
    c.ang_cost_func_type = 3
    c.use_cost_hess = True
    for f in ("plan_for_aero", "plan_for_gg", "plan_for_srp", "plan_for_prop",
              "plan_for_gendist", "plan_for_resdipole"):
        setattr(ps.disturbances, f, False)
    ps.passes[0].reg.reg_init = 1e-6
    ps.passes[0].reg.reg_max = 1e10
    ps.passes[0].reg.reg_scale = 10.0
    ps.passes[0].reg.use_dynamics_hess = False
    ps.passes[0].reg.use_constraint_hess = False
    ps.passes[0].linesearch.max_iters = 24
    ps.passes[0].linesearch.beta1 = 1e-10
    ps.passes[0].linesearch.beta2 = 5000.0
    ps.tvlqr.tvlqr_len = tvlqr_len
    ps.tvlqr.tvlqr_overlap = tvlqr_overlap
    return ps


def _solve(dt=10.0, tf=200.0, tvlqr_len=SINGLE_CHUNK_LEN, tvlqr_overlap=0.0):
    ps = _make_settings(dt, tvlqr_len, tvlqr_overlap)
    sat = create_satellite_rw(ps)
    jtime = np.array([0.22, 0.22 + tf / SEC_PER_CENTURY])
    r2 = np.sqrt(2.0) / 2.0
    qgoal = np.array([[r2, r2], [0.0, 0.0], [0.0, 0.0], [r2, r2]])
    boresight = np.array([[1.0, 1.0], [0.0, 0.0], [0.0, 0.0]])
    x0 = np.hstack(([-0.01, 0.02, 0.03], [1.0, 0.0, 0.0, 0.0], [0.0, 0.0, 0.0]))
    r0 = np.array([7000e3, 0.0, 0.0])
    v0 = np.array([0.0, 7.5e3, 0.0])
    ok, X, U, K = saltro_py.trajOpt(ps, sat, x0, r0, v0, jtime, qgoal, boresight)
    return dict(ok=ok, X=X, U=U, K=K, ps=ps, sat=sat, x0=x0, r0=r0, v0=v0, dt=dt)


def _gain_list(K, n_red, N):
    """Split the (input_dim, n_red*N) column-stacked K into N per-step blocks."""
    return [K[:, k * n_red:(k + 1) * n_red] for k in range(N)]


def _reduced_error(x_cur, x_plan, nRW):
    """Reduced-state error [dw, MRP(quatError), dh] -- the forward-pass convention."""
    dz = np.zeros(6 + nRW)
    dz[0:3] = x_cur[0:3] - x_plan[0:3]
    q_err = np.array(saltro_py.quatError(x_plan[3:7], x_cur[3:7]))
    dz[3:6] = np.array(saltro_py.quatToMRP(q_err))
    for i in range(nRW):
        dz[6 + i] = x_cur[7 + i] - x_plan[7 + i]
    return dz


def _rk4_step(sat, x, u, dist, R, B, S, V, rho, dt):
    def f(xx):
        return np.array(sat.dynamics(xx, u, dist, R, B, S, V, int(rho)))
    k1 = f(x)
    k2 = f(x + 0.5 * dt * k1)
    k3 = f(x + 0.5 * dt * k2)
    k4 = f(x + dt * k3)
    xn = x + (dt / 6.0) * (k1 + 2.0 * k2 + 2.0 * k3 + k4)
    xn[3:7] = xn[3:7] / np.linalg.norm(xn[3:7])  # stay on the unit sphere
    return xn


def _rollout(sat, ps, X, U, Klist, x0_perturbed, R, B, S, V, rho, dt, use_feedback):
    """Roll the (possibly perturbed) state forward, optionally applying eq. 7.39.

    Returns the per-step reduced tracking error ||x_k - x_plan_k||.
    """
    nRW = sat.stateDim - 7
    N = X.shape[1]
    x = x0_perturbed.copy()
    errs = []
    for k in range(N - 1):
        u = U[:, k].copy()
        if use_feedback:
            u = u + Klist[k] @ _reduced_error(x, X[:, k], nRW)
        x = _rk4_step(sat, x, u, ps.disturbances,
                      R[:, k], B[:, k], S[:, k], V[:, k], rho[0, k], dt)
        errs.append(np.linalg.norm(_reduced_error(x, X[:, k + 1], nRW)))
    return np.array(errs)


# ---------------------------------------------------------------------------
# Structural validity
# ---------------------------------------------------------------------------
def test_gains_have_correct_shape_and_are_finite_nonzero():
    r = _solve()
    assert r["ok"]
    n_red = r["sat"].reducedStateDim
    nu = r["sat"].controlDim
    N = r["X"].shape[1]
    K = r["K"]
    assert K.shape == (nu, n_red * N)
    assert np.all(np.isfinite(K))
    assert np.any(K != 0.0)


def test_gains_are_deterministic():
    K1 = _solve()["K"]
    K2 = _solve()["K"]
    assert np.array_equal(K1, K2)


# ---------------------------------------------------------------------------
# No-op limit: a window covering the whole horizon == the un-chunked gains
# ---------------------------------------------------------------------------
def test_window_covering_horizon_is_a_single_chunk_noop():
    # Two zero-overlap windows that both exceed the trajectory duration produce
    # the identical stitched gains: each is a single chunk spanning the whole
    # horizon, so the window length cannot change anything.
    Ka = _solve(tvlqr_len=400.0, tvlqr_overlap=0.0)["K"]
    Kb = _solve(tvlqr_len=5000.0, tvlqr_overlap=0.0)["K"]
    assert np.array_equal(Ka, Kb)


# ---------------------------------------------------------------------------
# Shorter windows actually change the gains, but stay bounded (no stitch blowup)
# ---------------------------------------------------------------------------
def test_short_windows_change_gains_but_stay_bounded():
    K_full = _solve(tvlqr_len=SINGLE_CHUNK_LEN, tvlqr_overlap=0.0)["K"]
    K_chunked = _solve(tvlqr_len=30.0, tvlqr_overlap=10.0)["K"]
    assert K_full.shape == K_chunked.shape
    assert np.all(np.isfinite(K_chunked))
    # Chunking must do something...
    assert np.max(np.abs(K_chunked - K_full)) > 1e-6
    # ...without blowing up at chunk boundaries (local-window gains are no larger
    # than a small multiple of the full-horizon gains).
    assert np.max(np.abs(K_chunked)) <= 5.0 * np.max(np.abs(K_full))


# ---------------------------------------------------------------------------
# The gains' purpose (eq. 7.39): feedback tracks the plan under a perturbation
# ---------------------------------------------------------------------------
@pytest.mark.parametrize("tvlqr_len, tvlqr_overlap", [(SINGLE_CHUNK_LEN, 0.0), (60.0, 15.0)])
def test_feedback_gains_track_plan_better_than_open_loop(tvlqr_len, tvlqr_overlap):
    r = _solve(tvlqr_len=tvlqr_len, tvlqr_overlap=tvlqr_overlap)
    assert r["ok"]
    sat, ps, X, U, dt = r["sat"], r["ps"], r["X"], r["U"], r["dt"]
    n_red = sat.reducedStateDim
    N = X.shape[1]
    Klist = _gain_list(r["K"], n_red, N)

    jt_fine = 0.22 + np.arange(N) * (dt / SEC_PER_CENTURY)
    ok_orbit, R, V, B, S, rho = saltro_py.generate_orbit(
        r["r0"], r["v0"], jt_fine, 0, 0, 0, 0, 0)
    assert ok_orbit
    rho = rho.reshape(1, -1)

    # Perturb the initial state: ~1 deg about +z and a small rate error.
    x0p = r["x0"].copy()
    x0p[0:3] += np.array([0.002, -0.0015, 0.001])
    half = 0.5 * np.deg2rad(1.0)
    dq = np.array([np.cos(half), 0.0, 0.0, np.sin(half)])
    q0 = r["x0"][3:7]
    x0p[3:7] = np.array([
        q0[0] * dq[0] - q0[1] * dq[1] - q0[2] * dq[2] - q0[3] * dq[3],
        q0[0] * dq[1] + q0[1] * dq[0] + q0[2] * dq[3] - q0[3] * dq[2],
        q0[0] * dq[2] - q0[1] * dq[3] + q0[2] * dq[0] + q0[3] * dq[1],
        q0[0] * dq[3] + q0[1] * dq[2] - q0[2] * dq[1] + q0[3] * dq[0],
    ])

    err_open = _rollout(sat, ps, X, U, Klist, x0p, R, B, S, V, rho, dt, use_feedback=False)
    err_closed = _rollout(sat, ps, X, U, Klist, x0p, R, B, S, V, rho, dt, use_feedback=True)

    assert np.all(np.isfinite(err_open))
    assert np.all(np.isfinite(err_closed))
    # Feedback must contain the perturbation: final tracking error at least 2x
    # smaller than open loop (empirically ~35x for the full-window gains).
    assert err_closed[-1] < 0.5 * err_open[-1]

File: tests/unit/optimizer/test_tvlqr_gains.cpp
// C++ twin of tests/unit/optimizer/test_tvlqr_gains.py.
//
// Characterizes trajOpt's chunked TVLQR feedback gains (the K output of
// compute_gains_chunked): structural validity, the single-chunk no-op limit,
// that shorter windows change the gains while staying bounded, and the gains'
// purpose -- applying eq. 7.39 to a perturbed closed-loop rollout tracks the
// plan far better than open loop. See the .py twin for the full rationale.
#include <catch2/catch_test_macros.hpp>

#include <cmath>
#include <vector>

#include <Eigen/Dense>

#include <saltro/limits.h>
#include <saltro/math/mrp.h>
#include <saltro/optimizer/trajOpt.h>
#include <saltro/orbit_generation/generate_orbit.h>
#include <saltro/pybind/satellite.h>

using namespace saltro;

namespace {

constexpr double SEC_PER_CENTURY = 36525.0 * 86400.0;

// A window longer than the (200 s) trajectory below with zero overlap is a
// single chunk spanning the whole horizon (see the .py twin for why overlap
// must be zero, and why the magnitude is kept modest).
constexpr double SINGLE_CHUNK_LEN = 1000.0;

PlannerSettings makeSettings(double dt_seconds, double tvlqr_len, double tvlqr_overlap) {
	PlannerSettings s;
	s.init_traj.initcontroller = 2;
	s.num_passes = 1;
	s.passes[0].dt = dt_seconds;
	s.passes[0].ilqr.cost_tol = 1e-5;
	s.passes[0].ilqr.max_iters = 20;
	s.passes[0].auglag.max_outer_iters = 10;
	s.passes[0].auglag.constraint_tol = 1e-3;

	auto& c = s.passes[0].cost;
	c.angle = 1.0;
	c.ang_vel = 1e1;
	c.control_mult = 1.0;
	c.mtq_control_weight = 1e-2;
	c.rw_control_weight = 1.0;
	c.ang_cost_func_type = 3;
	c.use_cost_hess = true;

	s.disturbances.plan_for_aero = false;
	s.disturbances.plan_for_gg = false;
	s.disturbances.plan_for_srp = false;
	s.disturbances.plan_for_prop = false;
	s.disturbances.plan_for_gendist = false;
	s.disturbances.plan_for_resdipole = false;

	s.passes[0].reg.reg_init = 1e-6;
	s.passes[0].reg.reg_max = 1e10;
	s.passes[0].reg.reg_scale = 10.0;
	s.passes[0].reg.use_dynamics_hess = false;
	s.passes[0].reg.use_constraint_hess = false;

	s.passes[0].linesearch.max_iters = 24;
	s.passes[0].linesearch.beta1 = 1e-10;
	s.passes[0].linesearch.beta2 = 5000.0;

	s.tvlqr.tvlqr_len = tvlqr_len;
	s.tvlqr.tvlqr_overlap = tvlqr_overlap;
	return s;
}

Eigen::Matrix3d rwInertia() {
	Eigen::Matrix3d J = Eigen::Matrix3d::Zero();
	J(0, 0) = 0.067;
	J(1, 1) = 0.071;
	J(2, 2) = 0.069;
	return J;
}

// Satellite has a deleted copy/move ctor, so it must be built in place; this
// adds the three reaction wheels to an already-constructed satellite.
void configureRW(Satellite& satellite) {
	satellite.addRW(Eigen::Vector3d::UnitX(), 0.001, 1e-5, 0.0, 0.02);
	satellite.addRW(Eigen::Vector3d::UnitY(), 0.001, 1e-5, 0.0, 0.02);
	satellite.addRW(Eigen::Vector3d::UnitZ(), 0.001, 1e-5, 0.0, 0.02);
}

Satellite::VecX makeX0(const Satellite& satellite) {
	Satellite::VecX x0(satellite.stateDim());
	x0.setZero();
	x0.segment<3>(Satellite::AV_INDEX) = Eigen::Vector3d(-0.01, 0.02, 0.03);
	x0.segment<4>(Satellite::QUAT_INDEX) = Eigen::Vector4d(1.0, 0.0, 0.0, 0.0);
	return x0;
}

struct SolveResult {
	bool ok = false;
	int N = 0;
	Eigen::MatrixXd X;  // state_dim x N
	Eigen::MatrixXd U;  // input_dim x N
	Eigen::MatrixXd K;  // input_dim x (reduced_state_dim * N)
};

SolveResult solveRW(double dt_seconds, double tf_seconds, double tvlqr_len,
                    double tvlqr_overlap) {
	const PlannerSettings settings = makeSettings(dt_seconds, tvlqr_len, tvlqr_overlap);
	Satellite satellite(rwInertia(), settings);
	configureRW(satellite);
	const Satellite::VecX x0 = makeX0(satellite);

	const Eigen::Vector3d r0(7000e3, 0.0, 0.0);
	const Eigen::Vector3d v0(0.0, 7.5e3, 0.0);

	Eigen::VectorXd jtime(2);
	jtime(0) = 0.22;
	jtime(1) = 0.22 + tf_seconds / SEC_PER_CENTURY;

	Eigen::MatrixXd q_goal(4, 2);
	q_goal << std::sqrt(2.0) / 2.0, std::sqrt(2.0) / 2.0,
	          0.0, 0.0,
	          0.0, 0.0,
	          std::sqrt(2.0) / 2.0, std::sqrt(2.0) / 2.0;
	Eigen::MatrixXd boresight(3, 2);
	boresight << 1.0, 1.0, 0.0, 0.0, 0.0, 0.0;

	const int state_dim = satellite.stateDim();
	const int input_dim = satellite.controlDim();
	const int reduced_state_dim = satellite.reducedStateDim();

	Eigen::MatrixXd X = Eigen::MatrixXd::Zero(state_dim, limits::MAX_LENGTH_TRAJ);
	Eigen::MatrixXd U = Eigen::MatrixXd::Zero(input_dim, limits::MAX_LENGTH_TRAJ);
	Eigen::MatrixXd K =
	    Eigen::MatrixXd::Zero(input_dim, reduced_state_dim * limits::MAX_LENGTH_TRAJ);
	int N = static_cast<int>(jtime.size());

	SolveResult r;
	REQUIRE_NOTHROW(r.ok = optimizer::trajOpt(settings, satellite, x0, r0, v0, jtime,
	                                           q_goal, boresight,
	                                           Eigen::MatrixXd(0, 0),  // seed_X
	                                           Eigen::MatrixXd(0, 0),  // seed_U
	                                           X, U, K, state_dim,
	                                           input_dim, N));
	r.N = N;
	r.X = X.leftCols(N);
	r.U = U.leftCols(N);
	r.K = K.leftCols(reduced_state_dim * N);
	return r;
}

// Reduced-state error [dw, MRP(quatError), dh] -- the forward-pass convention.
Eigen::VectorXd reducedError(const Eigen::VectorXd& x_cur, const Eigen::VectorXd& x_plan,
                             int nRW) {
	Eigen::VectorXd dz = Eigen::VectorXd::Zero(6 + nRW);
	dz.head<3>() = x_cur.head<3>() - x_plan.head<3>();
	const Eigen::Vector4d q_err =
	    math::quatError(x_plan.segment<4>(3), x_cur.segment<4>(3));
	dz.segment<3>(3) = math::quatToMRP(q_err);
	for (int i = 0; i < nRW; ++i) {
		dz(6 + i) = x_cur(7 + i) - x_plan(7 + i);
	}
	return dz;
}

Eigen::VectorXd rk4Step(const Satellite& sat, const Eigen::VectorXd& x,
                        const Eigen::VectorXd& u, const DisturbanceConfig& dist,
                        const Eigen::Vector3d& R, const Eigen::Vector3d& B,
                        const Eigen::Vector3d& S, const Eigen::Vector3d& V, int rho,
                        double dt) {
	const auto f = [&](const Eigen::VectorXd& xx) {
		return sat.dynamics(xx, u, dist, R, B, S, V, rho);
	};
	const Eigen::VectorXd k1 = f(x);
	const Eigen::VectorXd k2 = f(x + 0.5 * dt * k1);
	const Eigen::VectorXd k3 = f(x + 0.5 * dt * k2);
	const Eigen::VectorXd k4 = f(x + dt * k3);
	Eigen::VectorXd xn = x + (dt / 6.0) * (k1 + 2.0 * k2 + 2.0 * k3 + k4);
	xn.segment<4>(3).normalize();  // stay on the unit sphere
	return xn;
}

}  // namespace

TEST_CASE("chunked TVLQR gains have correct shape and are finite/nonzero",
          "[optimizer][tvlqr][gains]") {
	const SolveResult r = solveRW(10.0, 200.0, SINGLE_CHUNK_LEN, 0.0);
	REQUIRE(r.ok);
	Satellite satellite(rwInertia(), makeSettings(10.0, SINGLE_CHUNK_LEN, 0.0));
	configureRW(satellite);
	REQUIRE(r.K.rows() == satellite.controlDim());
	REQUIRE(r.K.cols() == satellite.reducedStateDim() * r.N);
	REQUIRE(r.K.allFinite());
	REQUIRE(r.K.cwiseAbs().maxCoeff() > 0.0);
}

TEST_CASE("chunked TVLQR gains are deterministic", "[optimizer][tvlqr][gains]") {
	const SolveResult a = solveRW(10.0, 200.0, SINGLE_CHUNK_LEN, 0.0);
	const SolveResult b = solveRW(10.0, 200.0, SINGLE_CHUNK_LEN, 0.0);
	REQUIRE(a.K.rows() == b.K.rows());
	REQUIRE(a.K.cols() == b.K.cols());
	REQUIRE((a.K - b.K).cwiseAbs().maxCoeff() == 0.0);
}

TEST_CASE("a window covering the horizon is a single-chunk no-op",
          "[optimizer][tvlqr][gains]") {
	const SolveResult a = solveRW(10.0, 200.0, 400.0, 0.0);
	const SolveResult b = solveRW(10.0, 200.0, 5000.0, 0.0);
	REQUIRE(a.K.cols() == b.K.cols());
	REQUIRE((a.K - b.K).cwiseAbs().maxCoeff() == 0.0);
}

TEST_CASE("shorter windows change the gains but stay bounded",
          "[optimizer][tvlqr][gains]") {
	const SolveResult full = solveRW(10.0, 200.0, SINGLE_CHUNK_LEN, 0.0);
	const SolveResult chunked = solveRW(10.0, 200.0, 30.0, 10.0);
	REQUIRE(full.K.cols() == chunked.K.cols());
	REQUIRE(chunked.K.allFinite());
	REQUIRE((chunked.K - full.K).cwiseAbs().maxCoeff() > 1e-6);
	REQUIRE(chunked.K.cwiseAbs().maxCoeff() <= 5.0 * full.K.cwiseAbs().maxCoeff());
}

TEST_CASE("feedback gains track the plan better than open loop (eq. 7.39)",
          "[optimizer][tvlqr][gains]") {
	const double dt = 10.0;
	const double tf = 200.0;
	const PlannerSettings settings = makeSettings(dt, SINGLE_CHUNK_LEN, 0.0);
	Satellite satellite(rwInertia(), settings);
	configureRW(satellite);
	const int nRW = satellite.stateDim() - 7;
	const int n_red = satellite.reducedStateDim();

	const SolveResult r = solveRW(dt, tf, SINGLE_CHUNK_LEN, 0.0);
	REQUIRE(r.ok);
	const int N = r.N;

	// Orbit environment along the fine trajectory.
	const Eigen::Vector3d r0(7000e3, 0.0, 0.0);
	const Eigen::Vector3d v0(0.0, 7.5e3, 0.0);
	Eigen::Matrix<double, 1, limits::MAX_LENGTH_TRAJ> jtime_full;
	jtime_full.setZero();
	const double dt_centuries = dt / SEC_PER_CENTURY;
	for (int k = 0; k < N; ++k) {
		jtime_full(0, k) = 0.22 + k * dt_centuries;
	}
	Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, V, B, S;
	Eigen::Matrix<double, 1, limits::MAX_LENGTH_TRAJ> rho;
	R.setZero();
	V.setZero();
	B.setZero();
	S.setZero();
	rho.setZero();
	REQUIRE(orbits::generate_orbit(r0, v0, jtime_full, N, 0, 0, 0, 0, 0, R, V, B, S, rho));

	// Perturb the initial state: ~1 deg about +z and a small rate error.
	const Satellite::VecX x0 = makeX0(satellite);
	Eigen::VectorXd x0p = x0;
	x0p.head<3>() += Eigen::Vector3d(0.002, -0.0015, 0.001);
	const double half = 0.5 * (1.0 * M_PI / 180.0);
	const Eigen::Vector4d dq(std::cos(half), 0.0, 0.0, std::sin(half));
	const Eigen::Vector4d q0 = x0.segment<4>(3);
	Eigen::Vector4d qp;
	qp(0) = q0(0) * dq(0) - q0(1) * dq(1) - q0(2) * dq(2) - q0(3) * dq(3);
	qp(1) = q0(0) * dq(1) + q0(1) * dq(0) + q0(2) * dq(3) - q0(3) * dq(2);
	qp(2) = q0(0) * dq(2) - q0(1) * dq(3) + q0(2) * dq(0) + q0(3) * dq(1);
	qp(3) = q0(0) * dq(3) + q0(1) * dq(2) - q0(2) * dq(1) + q0(3) * dq(0);
	x0p.segment<4>(3) = qp;

	const auto rollout = [&](bool use_feedback) {
		Eigen::VectorXd x = x0p;
		double final_err = 0.0;
		for (int k = 0; k < N - 1; ++k) {
			Eigen::VectorXd u = r.U.col(k);
			if (use_feedback) {
				const Eigen::MatrixXd K_k = r.K.block(0, k * n_red, r.K.rows(), n_red);
				u += K_k * reducedError(x, r.X.col(k), nRW);
			}
			x = rk4Step(satellite, x, u, settings.disturbances, R.col(k), B.col(k),
			            S.col(k), V.col(k), static_cast<int>(rho(0, k)), dt);
			final_err = reducedError(x, r.X.col(k + 1), nRW).norm();
		}
		return final_err;
	};

	const double err_open = rollout(false);
	const double err_closed = rollout(true);
	REQUIRE(std::isfinite(err_open));
	REQUIRE(std::isfinite(err_closed));
	REQUIRE(err_closed < 0.5 * err_open);
}
